epsie/proposals/birthdeath.py

```diff
@@ -0,0 +1,315 @@
+        for prop in self.all_proposals:
+            prop.bit_generator = self.bit_generator
+
+        self._last_prop = None
+
+
+    @property
+    def symmetric(self):
+        return self._symmetric
+
+    def logpdf(self, xi, givenx):
+        lp = 0.0
+        for prop in self.all_proposals:
+            if prop.active:
+                lp += prop.logpdf({p: xi[p] for p in prop.parameters},
+                                  {p: givenx[p] for p in prop.parameters})
+        return lp
+
+    def update(self, chain):
+        # update each of the proposals
+        acc = chain.acceptance[-1]['accepted']
+        if acc and (self._last_prop is not None):
+            self._last_prop.active = not self._last_prop.active
+
+        self._last_prop = None
+
+        # turn this back on later but need to check what happens to adaptation
+        # if proposals are getting turned off and on. Do adaptation only on
+        # proposals that just did a within-model MCMC step?
+        # Also what we could do is introduce an adaption of the mean of the
+        # poisson distribution that decides where to jump next.
+
+#        for prop in self.all_proposals:
+#            prop.update(chain)
+
+    @property
+    def prior_dist(self):
+        return self._prior_dist
+
+    @prior_dist.setter
+    def prior_dist(self, prior_dist):
+        try:
+            self._prior_dist = {p: prior_dist[p]
+                                for p in self.update_parameters}
+        except KeyError:
+            raise ValueError('provide a value for each parameter')
+
+    def jump(self, fromx):
+        givenk = fromx[self.model_parameter]
+        newk = self.bd_proposal.jump({self.model_parameter: givenk})
+        # unpack it
+        newk = newk[self.model_parameter]
+
+        inactive_props = list()
+        active_props = list()
+        for prop in self.update_proposals:
+            if prop.active:
+                active_props.append(prop)
+            else:
+                inactive_props.append(prop)
+#        print('active probs: ', len(active_props))
+#        print('inactive probs: ', len(inactive_props))
+#        print('-------------------')
+        out = {}
+        #print('fromx: ', fromx)
+        if newk != givenk:
+            if newk > givenk:
+#                print('tryng birth')
+                # pick which proposal to turn on
+                i = random.choice(range(len(inactive_props)))
+                birth_prop = inactive_props.pop(i)
+                # pick up the birth signal from the prior
+                out.update({p: self.prior_dist[p].rvs()
+                            for p in birth_prop.parameters})
+                self._last_prop = birth_prop
+
+            else:
+#                print('trying death')
+                # pick which proposal to turn off
+                i = random.choice(range(len(active_props)))
+                death_prop = active_props.pop(i)
```

Comment:
And here, toggle `death_prop.active = False` and add it to the list of `inactive_props`.

(Doing that can cause a proposal that was just birthed to be killed; is that ok?)